### ISSUE TITLE
Bivariant method check fails for index-signature generic type

### ISSUE THREAD

**Issue by ferdaber** [microsoft/TypeScript#47148]
# Bug Report
When instantiating a type with a broad index signature, and assigning it to another type with a more specific index signature whose properties are still assignable, bivariant method checking fails under a misleading error. This seems to only happen when the method under question uses a deferred keyof type query:

### 🔎 Search Terms
bivariant method function index signature keyof

### 🕗 Version & Regression Information
At least version 3.5

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about method bivariance checks

### ⏯ Playground Link
[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.3&ssl=26&ssc=67&pln=1&pc=1#code/C4TwDgpgBA0hIGcDyAzAKuCAeNaCGATgOYTAA0UaAwgPYB2CwBeAlneZXCFBAB7AQ6AEwRQA1vBopK+YqSgBecZOm5CJYAD5FneD36CRURgTZEoAWABQUW1AD8xpmf0DhotF0s27vxyZdrX18ALhl1UgBtT3gAXVdDD1oGJlZ2b2DMh11uIKzQqDoIADcIAgy7MKLSgmtrIQgAYwAbQmgWvARRWgBbMHpBYBwE9ygAJSaaAiEsALoiCgAjGhpmiDw6KAAfJ1N5zW0AbzzbRvoTAFdG4CmAChowYBZzsLQAShOoG6IiNduJEBhLjIdCYHBLFZrDaaN5hYo0FhCT4aNB8YD-eBA+AgjCQcG7MwwsJzIjWAC+dSsDQ6BGgxUIUBIRVMjV6-SK7AA6ixgAALZImNLAACCCAAIhAUHgLs1gGE2QN2FgJmdprNnPMIat1psdiSDvUmq1aVB6eUmWUWKyaH1FcBuXzhXQQKKJVKZXKoAqOUMNiBNNYzikoGxGBsnngBEJvYMdEUAO5em3swa3Y4+WwsBAACURDToYSlzQQEDInyzACEWERC3hi6XPj0IF08CQwgBybMQZrNGhQeNTZpCdvkj5WC0smNcnn886pNgi8WS6WynSh4DhliRiDR5N2gDcUAA9EevpgeAQCFMoAAmADMN5v3hPlHP7anQ0OIZzecEtfrh6VtW-4loeTYtm2BLzIeZKaO236FDQwBQJ0CDVnQeCLGsXx9qAkBQO+e4+sqkxqiSFAktsUDLNq0JwQAdNYL62Li0DtiqUwzORUHmDsNFQnQcEIT0WZofMXy8tAKCrL28YuGAV6QAQTzNlAKBXj0Z74e2X5ZrmQj5iBECAQgVY1mpdagVA4EIK2EDEhqRAwe2YR6b+dAUEBCzWc2tkkJSE5Wh+Dq8k6LpLu6q5KOum7bruto+oeL4DgQYgIEAA)

### 💻 Code
```ts
type KeysOfType<TTarget, TConstraint, TKey extends keyof TTarget = keyof TTarget> = TKey extends string 
    ? string extends TKey 
        ? string 
        : TTarget[TKey] extends TConstraint 
            ? TKey 
            : never 
    : never
    

declare class Component<T extends Record<string, boolean | string>> {
    constructor(options: T)
    toggle(key: KeysOfType<T, boolean>): void
    getText(key: KeysOfType<T, string>): string
}

declare var genericComponentWithConstraintAsDefault: Component<Record<string, boolean | string>>
declare var genericComponentWithAnyAsDefault: Component<any>
const instantiatedComponent = new Component({
    isHidden: false,
    isBig: false,
    message: 'Hello world'
})
genericComponentWithConstraintAsDefault = instantiatedComponent; // type error 2322 
// Type 'Component<{ isHidden: false; isBig: false; message: string; }>' is not assignable to type 'Component<Record<string, string | boolean>>'.
//   Type 'Record<string, string | boolean>' is missing the following properties from type '{ isHidden: false; isBig: false; message: string; }': isHidden, isBig, message

genericComponentWithAnyAsDefault = instantiatedComponent; // works
```

### 🙁 Actual behavior
The assignment of `genericComponentWithConstraintAsDefault` fails with the following error:
```
Type 'Component<{ isHidden: false; isBig: false; message: string; }>' is not assignable to type 'Component<Record<string, string | boolean>>'.
  Type 'Record<string, string | boolean>' is missing the following properties from type '{ isHidden: false; isBig: false; message: string; }': isHidden, isBig, message
```

### 🙂 Expected behavior
The assignment of `genericComponentWithConstraintAsDefault` does not fail. Note that when using `any` as the generic parameter, the assignment works. Also when removing the `string extends TKey ?` clause (which checks for wide type instantiations) in `KeysOfType`, it also removes the error, though I kind of need it to prevent the method parameters from being typed as `never`.

**Comment by RyanCavanaugh** [comment]
What's going on here is that we measure `Component` to be invariant, thus there's no need to check the structural relation between the two definitions of `toggle` since bivariance can't fix an invariant situation.

Similar demonstration + proof of should-be-assignable
```ts
// Appears invariant, actually covariant
type Inv<T> = {
    someValue: T[keyof T];
};

declare class Component<T> {
    constructor(options: T)
    toggle(key: Inv<T>): void
}

declare var genericComponentWithConstraintAsDefault: Component<Record<string, boolean | string>>
const instantiatedComponent = new Component({
    isHidden: false,
    isBig: false
});

instantiatedComponent.toggle({ someValue: true })
genericComponentWithConstraintAsDefault = instantiatedComponent; // type error 2322

// This assignment is OK -- methods are bivariantly related
const p: Parameters<(typeof genericComponentWithConstraintAsDefault)["toggle"]>[0] = null as any as Parameters<(typeof genericComponentWithConstraintAsDefault)["toggle"]>[0];
```

*However*, I think the example as given is actually a correct error, at least given the reasonable limitations of our analysis. `KeysOfType` really is invariant, because a covariant substitution of `TTarget` could produce a wider result set. *By construction* this situation can't happen with `Component` because of its constraint, but we have no way to reason about that fact.

**Comment by ferdaber** [comment]
When does the compiler determine a type to be invariant for assignability checks? I only know so far of enums and private fields but this one is new to me.

**Comment by RyanCavanaugh** [comment]
This is a different thing. Generic types are always "measured" in terms of their variance relationship to their type parameter(s). `Component` is invariant in `T` because it has both co- and contra-variant usages of `T`. This enables much faster relational checks since you can check `T<S> => T<U>` by checking `S => U`, `U => S`, or `U == S` depending on what that measurement produced.

**Comment by ferdaber** [comment]
I see, hence why the `any` substitution works. Thanks for the explanation!
